### PR TITLE
Add shader image query instructions

### PR DESCRIPTION
**Pull request by khyperia**
Ticks some more boxes off in #204

### CODE REVIEW

File: crates/rustc_codegen_spirv/src/builder/spirv_asm.rs [hunk 1]
```diff
@@ -1301,6 +1302,128 @@ impl<'cx, 'tcx> Builder<'cx, 'tcx> {
...
+                    Dim::Dim1D | Dim::Dim2D | Dim::Dim3D | Dim::DimCube => {
+                        multisampled == 1 || sampled == 0 || sampled == 2
+                    }
+                    Dim::DimBuffer | Dim::DimRect => true,
+                    Dim::DimSubpassData => false,
+                },
+                _ => true,
+            }
+        }
+        fn is_valid_query_size_lod(ty: &SpirvType) -> bool {
```
Comment (XAMPPRocky) [inline crates/rustc_codegen_spirv/src/builder/spirv_asm.rs:1355]:
```suggestion

        fn is_valid_query_size_lod(ty: &SpirvType) -> bool {
```

File: crates/rustc_codegen_spirv/src/builder/spirv_asm.rs [hunk 2]
```diff
@@ -1301,6 +1302,128 @@ impl<'cx, 'tcx> Builder<'cx, 'tcx> {
...
+                    // bad going on, spirv-val will catch it.
+                    return None;
+                }
+            };
+            match builder.lookup_type(defining_inst.result_type.unwrap()) {
+                SpirvType::SampledImage { image_type } => Some(builder.lookup_type(image_type)),
+                ty => Some(ty),
+            }
+        }
+        fn is_valid_query_size(ty: &SpirvType) -> bool {
```
Comment (XAMPPRocky) [inline crates/rustc_codegen_spirv/src/builder/spirv_asm.rs:1338]:
```suggestion

        fn is_valid_query_size(ty: &SpirvType) -> bool {
```

File: crates/spirv-std/src/image.rs [hunk 3]
```diff
@@ -659,6 +659,185 @@ impl<
...
+    /// Query the dimensions of Image, with no level of detail.
+    ///
+    /// Note: Const generics aren't able to reason about the constraints on this function (yet),
+    /// and so are enforced by the compiler. The constraints are:
+    ///
+    /// The image's dimension must be 1D, 2D, 3D, Buffer, Rect, or Cube. If dimension is 1D, 2D,
+    /// 3D, or Cube, it must have *either* multisampled be true, *or* sampled of Unknown or No.
+    #[crate::macros::gpu_only]
+    #[doc(alias = "OpImageQuerySize")]
+    pub fn query_size<Size: ImageCoordinate<u32, DIM, ARRAYED> + Default>(&self) -> Size {
```
Comment (XAMPPRocky) [inline crates/spirv-std/src/image.rs:743]: `ImageCoordinate` itself could probably require all implementations to also have Default as the only valid image coordinate types are vectors and scalars all of which impl `Default`.

File: tests/ui/image/query/query_levels_err.stderr [hunk 4]
```diff
@@ -0,0 +1,15 @@
+error: OpImageQueryLevels's image has a dimension of DimRect
+   --> /home/khyperia/me/rust-gpu/crates/spirv-std/src/image.rs:684:13
```
Comment (XAMPPRocky) [inline tests/ui/image/query/query_levels_err.stderr:2]: Seems incorrect that it has your personal details in the path.

File: tests/ui/image/query/query_lod_err.stderr [hunk 5]
```diff
@@ -0,0 +1,16 @@
+error: OpImageQueryLod's image has a dimension of DimRect
+   --> /home/khyperia/me/rust-gpu/crates/spirv-std/src/image.rs:717:13
```
Comment (XAMPPRocky) [inline tests/ui/image/query/query_lod_err.stderr:2]: Ditto

**Review by XAMPPRocky — APPROVED** [review]

**Review by XAMPPRocky — APPROVED** [review]

**Review by XAMPPRocky — APPROVED** [review]

**Review by eddyb — APPROVED** [review]

**Pull request merged**